gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java

```diff
@@ -0,0 +1,157 @@
+
+    protected DsEventStorage(DsRecordStorage<EventId> storage) {
+        super(storage);
+    }
+
+    @Override
+    protected Map<EventId, EntityRecord> readRecords(EventStreamQuery query) {
+        final StructuredQuery<Entity> entityQuery = buildQuery(query);
+        final DatastoreWrapper datastore = getDelegateStorage().getDatastore();
+        final List<Entity> readEntities = datastore.read(entityQuery);
+        if (readEntities.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        final Map<EventId, EntityRecord> result = processResults(readEntities);
+        return result;
+    }
+
+    private static StructuredQuery<Entity> buildQuery(EventStreamQuery sourceQuery) {
+        final Kind kind = Kind.of(TypeName.of(Event.class));
+        final Filter filter = buildFilter(sourceQuery);
+        final StructuredQuery.Builder<Entity> entityQuery = Query.newEntityQueryBuilder()
+                                                                 .setKind(kind.getValue());
+        if (filter == null) {
+            return entityQuery.build();
+        } else {
+            return entityQuery.setFilter(filter)
+                              .build();
+        }
+    }
+
+    @Nullable
+    private static Filter buildFilter(EventStreamQuery query) {
+        if (!query.hasBefore() && !query.hasAfter()) {
+            return null;
+        }
+        final DatastoreColumnType<Timestamp, DateTime> columnType = getTimestampColumnType();
+        Filter after = null;
+        if (query.hasAfter()) {
+            after = afterFilter(query.getAfter(), columnType);
+        }
+        Filter before = null;
+        if (query.hasBefore()) {
+            before = beforeFilter(query.getBefore(), columnType);
+        }
+        if (after != null && before != null) {
+            return and(before, after);
+        } else {
+            return after == null
+                   ? before
+                   : after;
+        }
+    }
+
+    private static Filter afterFilter(Timestamp after,
+                                      DatastoreColumnType<Timestamp, DateTime> columnType) {
+        final Value<?> value = columnType.toValue(columnType.convertColumnValue(after));
+        final Filter startTimeFilter = gt(EventEntity.CREATED_TIME_COLUMN, value);
+        return startTimeFilter;
+    }
+
+    private static Filter beforeFilter(Timestamp before,
+                                       DatastoreColumnType<Timestamp, DateTime> columnType) {
+        final Value<?> value = columnType.toValue(columnType.convertColumnValue(before));
+        final Filter startTimeFilter = lt(EventEntity.CREATED_TIME_COLUMN, value);
+        return startTimeFilter;
+    }
+
+    private static DatastoreColumnType<Timestamp, DateTime> getTimestampColumnType() {
+        final Class<EventEntity> eventEntityClass = EventEntity.class;
+        final Method getter;
+        try {
+            getter = eventEntityClass.getMethod("getCreated");
```

Comment:
This portion requires proper and detailed documentation in the method Javadoc. Please do, so we do not forget why it is done so.